File: content/problems/1048-longest-string-chain/solutions/delete-one-predecessor-dp/Solution.java
class Solution {
    public int longestStrChain(String[] words){
        Arrays.sort(words,Comparator.comparingInt(String::length));
        Map<String,Integer> best=new HashMap<>();
        int answer=1;
        for(String word:words){
            int length=1;
            for(int i=0;i<word.length();i++){
                String predecessor=word.substring(0,i)+word.substring(i+1);
                length=Math.max(length,best.getOrDefault(predecessor,0)+1);
            }
            best.put(word,length);
            answer=Math.max(answer,length);
        }
        return answer;
    }
}

File: content/problems/1048-longest-string-chain/solutions/delete-one-predecessor-dp/Solution.kt
class Solution {
    fun longestStrChain(words:Array<String>):Int{
        words.sortBy{
            it.length
        };
        val best=mutableMapOf<String,Int>();
        var answer=1;
        for(word in words){
            var length=1;
            for(i in word.indices){
                val predecessor=word.removeRange(i,i+1);
                length=maxOf(length,(best[predecessor]?:0)+1)
            };
            best[word]=length;
            answer=maxOf(answer,length)
        };
        return answer
    }
}
